### ISSUE TITLE
Proposed signature change for `tch::Tensor::grad`

### ISSUE THREAD

**Issue by antoniprzybylik** [LaurentMazare/tch-rs#946]
In PyTorch's python API calling `grad` not always returns a tensor. It may return `None` if the gradient is uninitialized:

```python
>>> import torch
>>> x = torch.tensor(1., requires_grad=True)
>>> print(type(x.grad))
<class 'NoneType'>
```

In **tch-rs** calling `grad` always returns a tensor:

```rust
pub fn grad(&self) -> Tensor
```

If the gradient is uninitialized then calling `grad` returns an undefined tensor. Attempts to perform operations on this undefined tensor may result in undefined behaviour. For example execution of this program:

```rust
fn main() {
    let x = tch::Tensor::from(1.).set_requires_grad(true);
    println!("x.grad(): {:#?}", x.grad());

    // This line will cause panic.
    let y = x.grad().copy();
}
```

Gives the following result:
```rust
x.grad(): Tensor[Undefined]

thread 'main' panicked at /home/antoni/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tch-0.19.0/src/wrappers/tensor_generated.rs:19318:29:
called `Result::unwrap()` on an `Err` value: Torch("tensor does not have a device\nException raised from device_default at /build/python-pytorch/src/pytorch-opt/c10/core/TensorImpl.h:1265 (most recent call first):\nframe #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0xa3 (0x7e614e5f6ad3 in /usr/lib/libc10.so)\nframe #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x86 (0x7e614e5953b2 in /usr/lib/libc10.so)\nframe #2: c10::KernelFunction::KernelFunction(std::unique_ptr<c10::OperatorKernel, std::default_delete<c10::OperatorKernel> >, void (*)(c10::OperatorKernel*, c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*), void*, void*) + 0 (0x7e6141df8a30 in /usr/lib/libtorch_cpu.so)\nframe #3: at::native::zeros_like(at::Tensor const&, std::optional<c10::ScalarType>, std::optional<c10::Layout>, std::optional<c10::Device>, std::optional<bool>, std::optional<c10::MemoryFormat>) + 0x12a (0x7e61425ddd4a in /usr/lib/libtorch_cpu.so)\nframe #4: <unknown function> + 0x28a637c (0x7e61434a637c in /usr/lib/libtorch_cpu.so)\nframe #5: at::_ops::zeros_like::call(at::Tensor const&, std::optional<c10::ScalarType>, std::optional<c10::Layout>, std::optional<c10::Device>, std::optional<bool>, std::optional<c10::MemoryFormat>) + 0x22f (0x7e61430d7d2f in /usr/lib/libtorch_cpu.so)\nframe #6: <unknown function> + 0x32730 (0x57da25f22730 in target/debug/blad-gradientu)\nframe #7: <unknown function> + 0x327dc (0x57da25f227dc in target/debug/blad-gradientu)\nframe #8: <unknown function> + 0x3301c (0x57da25f2301c in target/debug/blad-gradientu)\nframe #9: <unknown function> + 0x1399a (0x57da25f0399a in target/debug/blad-gradientu)\nframe #10: <unknown function> + 0x13cfa (0x57da25f03cfa in target/debug/blad-gradientu)\nframe #11: <unknown function> + 0x14ad6 (0x57da25f04ad6 in target/debug/blad-gradientu)\nframe #12: <unknown function> + 0x11c8a (0x57da25f01c8a in target/debug/blad-gradientu)\nframe #13: <unknown function> + 0x1110b (0x57da25f0110b in target/debug/blad-gradientu)\nframe #14: <unknown function> + 0x11a1e (0x57da25f01a1e in target/debug/blad-gradientu)\nframe #15: <unknown function> + 0x11e31 (0x57da25f01e31 in target/debug/blad-gradientu)\nframe #16: <unknown function> + 0x1f8a07 (0x57da260e8a07 in target/debug/blad-gradientu)\nframe #17: <unknown function> + 0x11e0a (0x57da25f01e0a in target/debug/blad-gradientu)\nframe #18: <unknown function> + 0x11d6e (0x57da25f01d6e in target/debug/blad-gradientu)\nframe #19: <unknown function> + 0x27488 (0x7e614093d488 in /usr/lib/libc.so.6)\nframe #20: __libc_start_main + 0x8c (0x7e614093d54c in /usr/lib/libc.so.6)\nframe #21: <unknown function> + 0x10ff5 (0x57da25f00ff5 in target/debug/blad-gradientu)\n")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I know that it is possible to check if a tensor is defined with `tch::Tensor::defined`, but I think that in safe Rust it should not be the programmer's responsibility to ensure that UB will not occur. There should not be any possibility of causing undefined behaviour in safe Rust.

I think that idiomatic Rust bindings should not return something that can be either valid result or undefined value. I think it would be more consistent with Rust conventions for `tch::Tensor::grad` to return `Option<Tensor>`.

The proposed signature of `tch::Tensor::grad` looks like that:
```rust
pub fn grad(&self) -> Option<Tensor>
```

**Comment by antoniprzybylik** [comment]
For this, `f_grad` can be used:
```rust
pub fn f_grad(&self) -> Result<Tensor, TchError>
```
